src/io/gridmanipulationio.py

```diff
@@ -0,0 +1,428 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
+
+
+import numpy as np
+import os
+import pandas as pd
+from types import SimpleNamespace
+from visualize import *
+# In[3]:
+
+
+class GridExtraction:
+    
+    def __init__(self, grid, flow):
+        self.grid = grid.grd
+        self.q = flow.q
+        
+        self.nb = grid.nb
+        self.ni = grid.ni
+        self.nj = grid.nj
+        self.nk = grid.nk
+        
+        self.mach = flow.mach
+        self.alpha = flow.alpha
+        self.rey = flow.rey
+        self.time = flow.time
+        
+        self.gd, self.fl = [], []
+            
+        self.df = pd.DataFrame(columns=['I_max', 'J_max', 'K_max'])
+
+        for n in range(int(self.nb)):
+            self.gd.append(self.grid[0:self.ni[n], 0:self.nj[n], 0:self.nk[n], :, n])
+            self.fl.append(self.q[0:self.ni[n], 0:self.nj[n], 0:self.nk[n], :, n])
+            
+            self.df.loc[len(self.df)] = ([self.ni[n], self.nj[n], self.nk[n]])
+            
+        print(self.df)
+        
+    def block_num(self, num):
+        self.num = num
+        self.v = Visualize(self.gd[self.num])
+
+    def extract(self, x_data=[None, None], y_data=[None, None], z_data=[None, None],
+                quantity=None, name=None,
+                representation='wireframe', scalar=False):
+        
+        self.x_data = x_data
+        self.y_data = y_data
+        self.z_data = z_data
+        
+        self.name = name
+        self.quantity = quantity
+        
+        if self.x_data == [None, None]: self.x_data = [0,self.ni[self.num]]
+        if self.y_data == [None, None]: self.y_data = [0,self.nj[self.num]]
+        if self.z_data == [None, None]: self.z_data = [0,self.nk[self.num]]
+
+        self.ext_grid = self.gd[self.num][self.x_data[0]:self.x_data[1], 
+                                          self.y_data[0]:self.y_data[1],
+                                          self.z_data[0]:self.z_data[1], :]
+        print("Grid data reading is successful! grid.shape: ", self.ext_grid.shape)
+        
+        self.ext_q = self.fl[self.num][self.x_data[0]:self.x_data[1],
+                                       self.y_data[0]:self.y_data[1],
+                                       self.z_data[0]:self.z_data[1], :]
+        print("\nFlow file read successfully! flow.shape: ",self.ext_q.shape)
+        
+        if self.name != None: self.v.add_parameter(self.quantity, self.name)
+        self.v.grid_extraction(self.x_data[0], self.x_data[1]-1,
+                          self.y_data[0], self.y_data[1]-1,
+                          self.z_data[0], self.z_data[1]-1,
+                          representation, scalar)
+        
```

Comment:
Go through the rest of the code and add comments at every line that needs an explantion